app/services/code_hosting/decorators/git_repo_decorator.rb

```diff
@@ -0,0 +1,100 @@
+require_relative "../../../shared/logging_module"
+require "securerandom"
+
+module FastlaneCI
+  # This module acts as a decorator to be added to any method that potentially will make changes
+  # into the ci-config repository directory.
+  #
+  #     @example
+  #       class Klass
+  #         extend FastlaneCI::GitRepoDecorator
+  #
+  #         def my_custom_function
+  #           method_that_reads_and_writes_files_of_ci_config_repo
+  #         end
+  #         pull_before(:my_custom_function, git_repo: my_git_repo)
+  #         commit_after(:my_custom_function, git_repo: my_git_repo)
+  #       end
+  #
+  # In this way every method that potentially can make changes over the ci-config repo is
+  # delayed until the ci-config repo is updated from remote, as it is our source of truth.
+  module GitRepoDecorator
+    # This decorator method is responsible of pulling and committing-pushing changes from the ci-config repo.
+    class << self
+
+      attr_reader :configuration_repo
+
+      def configuration_repository(ci_repo)
+        @configuration_repo = ci_repo
+      end
+
+      def pull_operation_mutex
+        @pull_operation_mutex ||= Mutex.new
+      end
+
+      def commit_operation_mutex
+        @commit_operation_mutex ||= Mutex.new
+      end
+
+      def included(base)
+        base.send(:extend, InstanceMethods)
+      end
+    end
+
+    # Add this as Instance Methods
+    module InstanceMethods
+      def pull_before(func_name, git_repo: GitRepoDecorator.configuration_repo)
+        new_name_for_old_function = "#{func_name}_old_pull_#{SecureRandom.uuid}".to_sym
```

Comment:
Sorry for the noob question: isn't there a better way in Ruby to achieve this? Seems kind of hacky and in-elegant

Reply:
This technique is similar to Rails' `alias_method_chain`. I _think_ we could achieve similar results using `Module#prepend` as mentioned here: https://www.justinweiss.com/articles/rails-5-module-number-prepend-and-the-end-of-alias-method-chain/

Reply:
My personal preference would be to try to use `Module#prepend`. If for whatever reason that is not possible here, I would prefer not to create methods with uuids in them. Something like `pull_before_#{func_name}` is totally reasonable. If that method already exists, perhaps we throw an exception, so we don't overwrite it.

Reply:
AFAIK, prepend just changes the call method chain, but for me to do the `prepend`, the module would need to have implemented the method that is being prepended like:
```ruby
module Decorator
   def method_prepended
      puts "I'm prepended"
   end
end
class Service
   prepend Decorator

   def method_prepended
      puts "I'm the last of the chain"
   end
end

# So this will end in:
service = Service.new
service.method_prepended # => "I'm prepended" "I'm the last of the chain"
```

So `Decorator` *needs* to know the name of the method that will be prepended before it actually is defined by the class it will prepend. I don't know if I'm explaining myself right 😅